File: solutions/CSharpSolutions/CSharpSolutions/Problems/1TwoSum.cs
namespace CSharpSolutions.Problems;

public class _1TwoSum
{
    public static int[] TwoSumBrute(int[] nums, int target)
    {
        for (int i = 0; i < nums.Length - 1; i++)
            for (int j = i + 1; j < nums.Length; j++)
                if (nums[i] + nums[j] == target)
                    return [i, j];
        throw new ArgumentException("No solution found");
    }

    public static int[] TwoSumSort(int[] nums, int target)
    {
        var numsWithIndices = new (int value, int index)[nums.Length];
        for (int i = 0; i < nums.Length; i++)
            numsWithIndices[i] = (nums[i], i);
        Array.Sort(numsWithIndices);
        for (int i = 0; i < nums.Length; i++)
        {
            int desired = target - numsWithIndices[i].value;
            var comparer = Comparer<(int value, int index)>.Create((a, b) => a.value - b.value);
            int found = Array.BinarySearch(numsWithIndices, i + 1, nums.Length - i - 1, (desired, 0), comparer);
            if (found > 0) return [numsWithIndices[i].index, numsWithIndices[found].index];
        }
        throw new ArgumentException("No solution found");
    }

    public static int[] TwoSumHash(int[] nums, int target)
    {
        var processedNums = new Dictionary<int, int>();
        for (int i = 0; i < nums.Length; i++)
        {
            if (processedNums.TryGetValue(target - nums[i], out int j))
                return [j, i];
            processedNums[nums[i]] = i;
        }
        throw new ArgumentException("No solution found");
    }
}

File: solutions/typescript-solutions/src/problems/1-two-sum.ts
import { binarySearch, NumWithIndex } from "../utils.js";

export function twoSumBrute(nums: number[], target: number): number[] {
    for (let i = 0; i < nums.length - 1; i++)
        for (let j = i + 1; j < nums.length; j++)
            if (nums[i] + nums[j] == target)
                return [i, j];
    throw new Error('No solution found');
}

export function twoSumSort(nums: number[], target: number): number[] {
    const numsWithIndices: NumWithIndex[] = [];
    for (let i = 0; i < nums.length; i++)
        numsWithIndices.push({ value: nums[i], index: i });
    numsWithIndices.sort((a, b) => a.value - b.value);
    for (let i = 0; i < nums.length; i++) {
        const desired = target - numsWithIndices[i].value;
        const found = binarySearch(numsWithIndices, { value: desired, index: 0 }, i + 1, nums.length - 1, (a, b) => a.value - b.value);
        if (found > 0)
            return [numsWithIndices[i].index, numsWithIndices[found].index];
    }
    throw new Error('No solution found');
}

export function twoSumHash(nums: number[], target: number): number[] {
    const processedNums: Map<number, number> = new Map();
    for (let i = 0; i < nums.length; i++) {
        const j = processedNums.get(target - nums[i]);
        if (j != undefined)
            return [j, i]
        processedNums.set(nums[i], i)
    }
    throw new Error('No solution found');
}
